posts/2021-07-28-GATs-stabilization-push.md

```diff
@@ -0,0 +1,132 @@
+
+Where to start, where to start...
+
+Let's begin by saying: this is a very exciting post. Some people reading this will be overwhelmingly thrilled; some will have no idea what this feature is; others might be in disbelief. The [RFC] for this feature *did* get opened in April of *2016* (and merged about a year and a half later). In fact, this RFC even predates *const generics*. Don't let this fool you though: it is a powerful feature; and the reactions to the tracking issue on Github should maybe give you an idea of its popularity (it *is* the most upvoted issue on the Rust repository):
+![GATs reactions](https://i.imgur.com/6uOZbXj.png)
+
+So, what exactly *is* happening. Well, nearly four years after its RFC was merged, the `generic_associated_types` feature is no longer "incomplete."
+
+*crickets chirping*
+
+Wait...that's it?? Well, yes! I'll go into a bit of detail later in this blog post as to why this *is* a big deal. But, long story short, there have been a good amount of changes that have had to have been made to the compiler to get GATs to work. And, while there are still a few small remaining diagnostics issues, the feature is finally in a space that we feel comfortable making it no longer "incomplete".
+
+So, what does that mean? Well, all it *really* means is that when you use this feature on nightly, you'll no longer get the "`generic_associated_types` is incomplete" warning. However, the real reason this is a big deal: we want to stabilize this feature. But we need *your* help. We need you to test this feature; file issues for bugs or for potential diagnostics improvements.
+
+Without making promises that we aren't 100% sure we can keep, we have high hopes we can stabilize this feature within the next couple months. But, we want to make sure we aren't missing glaringly obvious bugs or flaws. And that's why we need *your* help.
+
+
+Okay. Phew. That's the main point of this post and the most exciting news. But as I said before, I think it's also reasonable for me to explain *what* this feature is, what you can do with it, and some of the background and how we got here.
+
+## So what are GATs?
+
+**Note: this will only be a brief overview. The [RFC] contains many more details.**
+
+GATs (generic associated types) were originally proposed in [RFC 1598][RFC]. Put simply, they allow you to define type, lifetime, or const generics on associated types. If you're familiar with languages that have "higher-kinded types", then you could calls GATs *type constructors on traits*.
+
+To jump into an example of GATs, here is a popular use case: a `LendingIterator` (formerly known as a `StreamingIterator`):
+
+```rust
+trait LendingIterator {
+    type Item<'a>;
+    fn next<'a>(&'a mut self) -> Option<Self::Item<'a>>;
+}
+```
+
+With this, you could create a line iterator over a `String` without cloning each line...
+```rust
+struct LineIterator { ... }
+impl LendingIterator for LineIterator {
+    type Item<'a> = &'a str;
```

Comment:
This is a bit of a strange example since the [Lines](https://doc.rust-lang.org/stable/std/str/struct.Lines.html) already exists. It probably won't be clear to those unfamiliar with GATs why this is different.

Reply:
maybe the example should be `<[T]>::windows_mut`:
```rust
struct WindowsMut<'a, T: 'a> {
    slice: &'a mut [T],
    start: usize,
    window_size: NonZeroUsize,
}

impl<'a, T: 'a> LendingIterator for WindowsMut<'a, T> {
    type Item<'b> = &'b mut [T];
    fn next<'b>(&'b mut self) -> Option<Self::Item<'b>> {
        let start = self.start;
        let retval = self.slice[start..].get_mut(..self.window_size.get())?;
        self.start = start + 1;
        Some(retval)
    }
}
```

Reply:
This could be good. We should include the error if you try to implement this with `Iterator`:

```
error[E0495]: cannot infer an appropriate lifetime for lifetime parameter in function call due to conflicting requirements
  --> src/lib.rs:12:22
   |
12 |         let retval = self.slice[start..].get_mut(..self.window_size.get())?;
   |                      ^^^^^^^^^^^^^^^^^^^
   |
note: first, the lifetime cannot outlive the anonymous lifetime defined on the method body at 10:13...
  --> src/lib.rs:10:13
   |
10 |     fn next(&mut self) -> Option<Self::Item> {
   |             ^^^^^^^^^
note: ...so that reference does not outlive borrowed content
  --> src/lib.rs:12:22
   |
12 |         let retval = self.slice[start..].get_mut(..self.window_size.get())?;
```

This shows that `self` is being borrowed for some lifetime and `Self::Item` is borrowed for the lifetime `'a`, but we have no way to tell the compiler how these lifetimes relate. With GATs we can explicitly say that `self` and `Self::Item` are borrowed for the same lifetime.

Reply:
I took the advice here and made the main example the `windows_mut`. I gave an example how you might try to implement it today, the error you get, and how you actually implement it with GATs. How does it look?

Reply:
looks good to me!